### PR TITLE
Add check for digipeater callsign and hbit to block 'Check address'

### PR DESCRIPTION
**Pull request by DL7NDR**
Adds check for digipeater callsign and hbit set.
In combination with 'telemetry forwarder' good to filter out decoded own frames during transmitting (near TCA).

### CODE REVIEW

**Comment by daniestevez** [comment]
First off, it is not necessary to open and close multiple issues form the same thing (#684, #685, #686, #687, #688, #689). For each of these people subscribed to the repository notifications receive an email. You can push additional commits to the same branch or do `git commit --amend` and `git push --force`.

File: python/check_address.py [hunk 1]
```diff
@@ -52,14 +53,42 @@ def handle_msg(self, msg_pmt):
         else:
             address = packet[7:14]
 
+        if len(packet) > 20:
```
Comment (daniestevez) [inline python/check_address.py:56]: What happens if `len(packet) <= 20`? It looks like `hbit` and `digi` are not defined, but they are read afterwards regardless.

File: python/check_address.py [hunk 2]
```diff
@@ -52,14 +53,42 @@ def handle_msg(self, msg_pmt):
...
+# digi callsign; and above: hbit  (most significant bit after digi callsign)
+# (set to 1 in case of digipeated message)
+            digi = [c >> 1 for c in packet[14:20]]
+
+# error handling
+            try:
+                digi = bytes(digi).decode('ascii').rstrip(' ')
+
+            except UnicodeDecodeError:
+                print("Not to ASCII convertable string detected.")
```
Comment (daniestevez) [inline python/check_address.py:68]: How can this possibly fail to decode as ASCII? It is a `bytes` object where all the bytes have the MSB equal to zero. Regardless, in the event there was a `UnicodeDecodeError` exception, `digi` would be undefined but potentially read below.

File: python/check_address.py [hunk 3]
```diff
@@ -52,14 +53,42 @@ def handle_msg(self, msg_pmt):
...
-        if (callsign != self.callsign
-                or (self.ssid is not None and ssid != self.ssid)):
-            # Incorrect address
-            self.message_port_pub(pmt.intern('fail'), msg_pmt)
-        else:
-            # Correct address
+
+# altered if sentence
+        if ((callsign == self.callsign
+                and (ssid == self.ssid if self.ssid is not None else True))
```
Comment (daniestevez) [inline python/check_address.py:88]: This condition is better written as `self.ssid is None or ssid == self.ssid`.

**Review by daniestevez — CHANGES_REQUESTED** [review]
This feature should be optional and default to the current behaviour. The changes cannot be merged in their present form, because they change this block's behaviour and introduce a new constructor parameter without a default value, which breaks API compatibility.

File: python/check_address.py (hunk 1 above)
Comment (DL7NDR) [reply python/check_address.py:56]: What happens if `len(packet) <= 20`?
->`hbit` and `digi` won't be defined.

Is it ok for you to initialize `hbit` and `digi` right before `len(packet) <= 20` by:
```
        hbit = 0
        digi = ""
```

Comment (daniestevez) [reply python/check_address.py:56]: I guess not. What's the code supposed to do in this case? If repeater address checking has been enabled but the message doesn't even contain repeater addresses because it is too short, shouldn't it go directly to the `fail` output?

File: python/check_address.py (hunk 2 above)
Comment (DL7NDR) [reply python/check_address.py:68]: You are right. Then we just remove it.

File: python/check_address.py (hunk 1 above)
Comment (DL7NDR) [reply python/check_address.py:56]: If `len(packet) <=20`, then hbit is 0 and there is no digipeater (digi = "").
In this case, the regular sender / destination callsign check should and will be processed anyway.

Comment (daniestevez) [reply python/check_address.py:56]: I had to read again the condition you wrote on lines 87-89 to understand what this feature is supposed to do. Let me rephrase: this feature will pass the frame to `ok` if either its destination address matches `address` or if the frame contains 3 addresses, the H bit is set (indicating that the frames has been repeated) and the repeater address (3rd address) matches `digicallsign`.

In that case I agree that just setting `hbit = 0` if `len(packet) <= 20` might be enough.

Comment (DL7NDR) [reply python/check_address.py:56]: > (...) if either its destination address matches `address` (...)
yes, destination or source (according to the chosen _direction_ "to" or "from")

Correct, that's the way I wanted it to work.

**Comment by daniestevez** [comment]
@DL7NDR please take note of my previous remark:

> The changes cannot be merged in their present form, because they change this block's behaviour and introduce a new constructor parameter without a default value, which breaks API compatibility.

More concretely, the `digicallsign` argument in the block's class constructor should have a default value (possibly `''` or `None`), the GRC YAML file should also default to that value, and the default value should achieve the same behaviour that the block currently has (i.e., only the source or destination callsign is checked).

File: grc/satellites_check_address.block.yml [hunk 4]
```diff
@@ -28,6 +32,18 @@ outputs:
 
 templates:
     imports: import satellites
-    make: satellites.check_address(${address}, ${direction})
+    make: satellites.check_address(${address}, ${direction}, ${digicallsign})
+
+documentation: |-
+    Enter an address (choose "from" for source or "to" for destination callsign) and/or enter a digipeater callsign.
+    The address callsign may have a SSID, the digipeater callsign must not have a SSID.
```
Comment (daniestevez) [inline grc/satellites_check_address.block.yml:39]: Can you describe accurately which frames are passed to the `ok` output? At the moment nothing in this documentation gives a clue as to whether the `or` in line 79 is an `or` or an `and`.

**Review by daniestevez — APPROVED** [review]
This looks good now. Approved and merging.

**Comment by daniestevez** [comment]
I merged too fast. I didn't realize this was actually 7 commits. I should have squashed. I'll do this manually now.
